### PR TITLE
Begin async tenant API FOLIO-2877

### PR DESCRIPTION
**Pull request by adamdickmeiss**
## Purpose
Async tenant API which we need because they may take long (much longer than many HTTP 
proxies would like).

## Approach
Update RAML .. so that RMB and Okapi can use the definition(s) and implement the interfaces.

### CODE REVIEW

**Review by dcrossleyau — APPROVED** [review]
I have verified the RAML/Schema with the new AMF tools. All happy.

File: schemas/tenantAttributes.schema [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{ "$schema": "http://json-schema.org/draft-04/schema#",
+  "id": "tenantAttributes.schema",
+  "title":  "Tenant job attributes",
+  "description": "This is the structure that's POSTed when a module is told about a tenant using it",
```
Comment (julianladisch) [inline schemas/tenantAttributes.schema]:
```suggestion
  "description": "Configuration how to install, upgrade or delete a module for a tenant",
```

File: ramls/tenant.raml [hunk 2]
```diff
@@ -2,68 +2,79 @@
...
-        description: "true or false indicator"
-        body:
-          text/plain:
-            example: "Text....."
-      500:
-        description: "Internal server error"
-        body:
-          text/plain:
-            example: "Internal server error, contact administrator"
-  delete:
```
Comment (julianladisch) [inline ramls/tenant.raml:56]: Since RMB 30.0.0 we store the module version in the database:
https://github.com/folio-org/raml-module-builder/blob/v30.0.0/domain-models-runtime/src/main/resources/templates/db_scripts/main.ftl#L210
Therefore there is no longer a need to require the module_from parameter.
We also don't need the module_to parameter because the module knowns its current version that we run (for creation or for upgrading to).

I suggest to keep both the POST and DELETE endpoint. They can be used without module_from and module_to parameters:
POST `/_/tenant` creates if there is no database schema yet.
POST `/_/tenant` upgrades if the the database schema already exists.
DELETE `/_/tenant` deletes the database schema.
It might be useful to have
GET `/_/tenant` that returns 404 if no database schema exists, otherwise returns "moduleId" and "dataModuleId". "dataModuleId" is optional and is only used for storage modules. If "moduleId" and "dataModuleId" differ then a POST `/_/tenant` is needed to upgrade the database.

Comment (adamdickmeiss) [reply ramls/tenant.raml:56]: module_to and module_from may be ignored at the moment.. But they are still passed to SQL scripts (if they are not actually in use, that code should be removed). They were in use at a point, so it might be in used by another implementation than just RMB. So the motivation for removing them because RMB may not use them is not really an argument. In particular any module that doesn't use storage will not know module_from

RMB in fact at the moment has weird semantics when a module is disabled (module_from given, but no module_to). It does nothing.. It most certainly does not disable the tenant. It should at least disable the tenant by some means.. Of course not purge data. This is what happens on `apt remove` in Debian.

As for the DELETE and keeping that. I thought it would be pointless. Mostly because it's not async. RMB might be fast for purge, but in the generic case. So that's why PURGE is just another Tenant Job. And the implementation only has to implement one POST method.

Remember that the initial design of the tenant API was CRUD for the Tenant. Now it's CRUD for the Tenant job "operation". That's also why both GET and DELETE would not be part of a new CRUD job API to get tenant or remove it. It is to get job status and delete job.

As for old tenant API: the GET was never used by Okapi. Not to mention that it is lying. If anything it should safe false if a Tenant is disabled. It doesn't. It only returns NOT FOUND if tenant is purged earlier. Again.. The existing design was CRUD for Tenant.

So to summarize I don't see a good argument for changing the proposed async API. What is missing though is a more fined tuned Tenant Job API to be able to support something ala 3-way commit (later).

**Comment by adamdickmeiss** [comment]
If we assume that a module can only work with one upgrade at a time, then we could make an API without a returned ID+Location. This would mean:

1. POST initiates the upgrade job.. Fails if one is already in progress (not completed).
2. GET gets the last/current job.
3. No DELETE anymore..

It would make it easier to use the API and also to implement it.

File: ramls/tenant.raml (hunk 2 above)
Comment (julianladisch) [reply ramls/tenant.raml:56]: If a module does not have any storage there is nothing it can migrate, and no need to know module_from.

**Review by julianladisch — APPROVED** [review]
